crates/mod-aws-sigv4/src/lib.rs

```diff
@@ -0,0 +1,315 @@
+
+    // Create payload hash
+    let payload_hash = sha256_hex(req.payload.as_bytes());
+
+    // Prepare headers - add required AWS headers
+    let mut headers = req.headers.clone();
+    headers.insert("host".to_string(), "".to_string()); // Will be set by caller
+    headers.insert("x-amz-date".to_string(), amz_date.clone());
+
+    if let Some(token) = &req.session_token {
+        headers.insert("x-amz-security-token".to_string(), token.clone());
+    }
+
+    // Add content hash header for some services
+    if req.service != "s3" {
+        headers.insert("x-amz-content-sha256".to_string(), payload_hash.clone());
+    }
+
+    // Create canonical request
+    let canonical_uri = create_canonical_uri(&req.uri);
+    let canonical_query_string = create_canonical_query_string(&req.query_params);
+    let (canonical_headers, signed_headers) = create_canonical_headers(&headers);
+
+    let canonical_request = format!(
+        "{}\n{}\n{}\n{}\n\n{}\n{}",
+        req.method,
+        canonical_uri,
+        canonical_query_string,
+        canonical_headers,
+        signed_headers,
+        payload_hash
+    );
+
+    // Create string to sign
+    let algorithm = "AWS4-HMAC-SHA256";
+    let credential_scope = format!("{}/{}/{}/aws4_request", date_stamp, req.region, req.service);
+    let canonical_request_hash = sha256_hex(canonical_request.as_bytes());
+
+    let string_to_sign = format!(
+        "{}\n{}\n{}\n{}",
+        algorithm, amz_date, credential_scope, canonical_request_hash
+    );
+
+    // Calculate signature
+    let signing_key = create_signing_key(&secret_key, &date_stamp, &req.region, &req.service);
+    let signature_bytes = hmac_sha256(&signing_key, string_to_sign.as_bytes());
+    let signature = HEXLOWER.encode(&signature_bytes);
+
+    // Create authorization header
+    let authorization = format!(
+        "{} Credential={}/{}, SignedHeaders={}, Signature={}",
+        algorithm, req.access_key, credential_scope, signed_headers, signature
+    );
+
+    Ok(SigV4Response {
+        authorization,
+        timestamp: amz_date,
+        canonical_request,
+        string_to_sign,
+        signature,
+    })
+}
+
+pub fn register(lua: &Lua) -> anyhow::Result<()> {
+    let aws_mod = get_or_create_sub_module(lua, "aws")?;
+
+    aws_mod.set(
+        "sign_v4",
+        lua.create_async_function(|lua, request: Value| async move {
+            let req: SigV4Request = from_lua_value(&lua, request)?;
+            let response = sign_request(req).await.map_err(any_err)?;
+
+            let result = lua.create_table()?;
+            result.set("authorization", response.authorization)?;
+            result.set("timestamp", response.timestamp)?;
+            result.set("signature", response.signature)?;
+            result.set("canonical_request", response.canonical_request)?;
+            result.set("string_to_sign", response.string_to_sign)?;
+
+            Ok(result)
```

Comment:
Can use the `Serialize` impl to take care of this boilerplate:

```suggestion
            lua.to_value(&result)
```

Reply:
Done – inside `register`, after calling `sign_request`, we now just call `lua.to_value(&response)` instead of manually constructing a Lua table. The field names and Lua-visible structure remain the same.